plugin/storage/cassandra/samplingstore/storage.go

```diff
@@ -0,0 +1,282 @@
+	probabilities model.ServiceOperationProbabilities,
+	qps model.ServiceOperationQPS,
+) error {
+	probabilitiesAndQPSStr := probabilitiesAndQPSToString(probabilities, qps)
+	query := s.session.Query(insertProbabilities, constBucket, gocql.TimeUUID(), hostname, probabilitiesAndQPSStr)
+	return s.metrics.probabilities.Exec(query, s.logger)
+}
+
+// GetLatestProbabilities implements samplingstore.Reader#GetLatestProbabilities.
+func (s *SamplingStore) GetLatestProbabilities() (model.ServiceOperationProbabilities, error) {
+	iter := s.session.Query(getLatestProbabilities).Iter()
+	var probabilitiesStr string
+	iter.Scan(&probabilitiesStr)
+	if err := iter.Close(); err != nil {
+		err = errors.Wrap(err, "Error reading probabilities from storage")
+		return nil, err
+	}
+	return s.stringToProbabilities(probabilitiesStr), nil
+}
+
+// GetProbabilitiesAndQPS implements samplingstore.Reader#GetProbabilitiesAndQPS.
+func (s *SamplingStore) GetProbabilitiesAndQPS(start, end time.Time) (map[string][]model.ServiceOperationData, error) {
+	iter := s.session.Query(getProbabilities, gocql.UUIDFromTime(start), gocql.UUIDFromTime(end)).Iter()
+	hostProbabilitiesAndQPS := make(map[string][]model.ServiceOperationData)
+	var probabilitiesAndQPSStr, host string
+	for iter.Scan(&probabilitiesAndQPSStr, &host) {
+		hostProbabilitiesAndQPS[host] = append(hostProbabilitiesAndQPS[host], s.stringToProbabilitiesAndQPS(probabilitiesAndQPSStr))
+	}
+	if err := iter.Close(); err != nil {
+		err = errors.Wrap(err, "Error reading probabilities and qps from storage")
+		return nil, err
+	}
+	return hostProbabilitiesAndQPS, nil
+}
+
+// This is random enough for storage purposes
+func generateRandomBucket() int64 {
+	return time.Now().UnixNano() % 10
+}
+
+func probabilitiesAndQPSToString(probabilities model.ServiceOperationProbabilities, qps model.ServiceOperationQPS) string {
+	var buf bytes.Buffer
+	writer := csv.NewWriter(&buf)
+	for svc, opProbabilities := range probabilities {
+		for op, probability := range opProbabilities {
+			opQPS := 0.0
+			if _, ok := qps[svc]; ok {
+				opQPS = qps[svc][op]
+			}
+			writer.Write([]string{svc, op, strconv.FormatFloat(probability, 'f', -1, 64),
+				strconv.FormatFloat(opQPS, 'f', -1, 64)})
+		}
+	}
+	writer.Flush()
+	return buf.String()
+}
+
+func (s *SamplingStore) stringToProbabilitiesAndQPS(probabilitiesAndQPSStr string) model.ServiceOperationData {
+	probabilitiesAndQPS := make(model.ServiceOperationData)
+	appendFunc := s.appendProbabilityAndQPS(probabilitiesAndQPS)
+	s.parseString(probabilitiesAndQPSStr, 4, appendFunc)
+	return probabilitiesAndQPS
+}
+
+func (s *SamplingStore) stringToProbabilities(probabilitiesStr string) model.ServiceOperationProbabilities {
+	probabilities := make(model.ServiceOperationProbabilities)
+	appendFunc := s.appendProbability(probabilities)
+	s.parseString(probabilitiesStr, 4, appendFunc)
+	return probabilities
+}
+
+func throughputToString(throughput []*model.Throughput) string {
```

Comment:
Given this is the only location it's ever used, I don't really see the need to do this. However I do like the encapsulation it will give us. So my answer is meh?